service/src/main/kotlin/app/cash/backfila/service/runner/BackfillRunner.kt

```diff
@@ -180,24 +187,38 @@ class BackfillRunner private constructor(
     failuresSinceSuccess = 0
   }
 
-  fun onRpcFailure() {
+  suspend fun onRpcFailure(
+    exception: Exception,
+    action: String,
+    elapsed: Duration
+  ) {
     // If there is an intermittent server issue, all the current batches will likely fail.
     // So to consider those as only one failure, only increment failure count if the backoff
     // finished.
     if (globalBackoff.backingOff()) {
       logger.info { "Ignoring rpc error because runner is already backing off ${logLabel()}" }
+      recordErrorEvent(exception, action, elapsed, backoffMs = null, paused = false)
       return
     }
+
     failuresSinceSuccess++
     if (failuresSinceSuccess > metadata.backoffSchedule.size) {
       logger.info {
         "Pausing backfill ${logLabel()} due to too many consecutive failures: $failuresSinceSuccess"
       }
       if (pauseBackfill()) {
         factory.slackHelper.runErrored(backfillRunId)
+
+        recordErrorEvent(exception, action, elapsed, backoffMs = null, paused = true)
       }
+
+      // Give the main loop a chance to run and exit.
```

Comment:
Can you explain why this is here?

Reply:
oh yeah this is a unrelated bugfix for a bug I ran into in development, cause it would fail instantly with a not-bound exception but because coroutines are cooperative scheduling, this coroutine kept executing and failing thousands of time and not giving other ones a chance to run

eg the BatchQueuer is supposed to be waiting on an RPC, but always instantly gets a binding exception. Since it's never waiting, BatchRunner can never run, neither can the "leaser" task in the BackfillRunner which is supposed to be what cancels the whole thing when `running = false`

I'll update the comment with more detail

Reply:
Ok sounds good.